src/coreclr/jit/lower.cpp

```diff
@@ -3777,12 +3777,47 @@ void Lowering::LowerRetStruct(GenTreeUnOp* ret)
 
         case GT_BLK:
         case GT_OBJ:
-            retVal->ChangeOper(GT_IND);
-            FALLTHROUGH;
         case GT_IND:
+        {
+            // Spill to a local if sizes don't match so we can avoid the "load more than requested"
+            // problem, e.g. struct size is 5 and we emit "ldr x0, [x1]"
+            unsigned realSize = retVal->OperIs(GT_OBJ) ? retVal->AsObj()->Size() : retVal->AsIndir()->Size();
+            CORINFO_CLASS_HANDLE structCls = NO_CLASS_HANDLE;
```

Comment:
```suggestion
            CORINFO_CLASS_HANDLE structCls = comp->info.compMethodInfo->args.retTypeClass;
```
(And corresponding change below)

We should always use a custom temp, even if we know the size, because we have `GT_BLK` indirs that will have the size but not the handle.